dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -233,23 +247,89 @@ class AsyncTunnelSender : public TunnelSender
     GRPCSendQueue<TrackedMppDataPacketPtr> queue;
 };
 
-/// LocalTunnelSender just provide readForLocal method to return one element one time
-/// LocalTunnelSender is owned by the associated ExchangeReceiver
+template <bool enable_fine_grained_shuffle>
 class LocalTunnelSender : public TunnelSender
 {
 public:
-    using Base = TunnelSender;
-    using Base::Base;
-    TrackedMppDataPacketPtr readForLocal();
+    LocalTunnelSender(
+        size_t source_index_,
+        LocalRequestHandler & local_request_handler_,
+        const LoggerPtr & log_,
+        MemoryTrackerPtr & memory_tracker_,
+        const String & tunnel_id_)
+        : TunnelSender(memory_tracker_, log_, tunnel_id_, nullptr)
+        , source_index(source_index_)
+        , local_request_handler(local_request_handler_)
+        , is_done(false)
+    {}
+
+    ~LocalTunnelSender() override
+    {
+        closeLocalTunnel(false, "");
+    }
+
+    bool push(TrackedMppDataPacketPtr && data) override
+    {
+        if (unlikely(checkPacketErr(data)))
+            return false;
+
+        // receiver_mem_tracker pointer will always be valid because ExchangeReceiverBase won't be destructed
+        // before all local tunnels are destructed so that the MPPTask which contains ExchangeReceiverBase and
+        // is responsible for deleting receiver_mem_tracker must be destroyed after these local tunnels.
+        data->switchMemTracker(local_request_handler.recv_mem_tracker);
+
+        auto res = local_request_handler.write<enable_fine_grained_shuffle>(source_index, data);
+
+        if (unlikely(!res))
+            closeLocalTunnel(true, "Push mpp packet failed at local tunnel");
+        return res;
+    }
+
+    void cancelWith(const String & reason) override
+    {
+        closeLocalTunnel(true, reason);
+    }
+
+    bool finish() override
+    {
+        closeLocalTunnel(false, "");
+        return true;
+    }
 
 private:
-    bool cancel_reason_sent = false;
+    bool checkPacketErr(TrackedMppDataPacketPtr & packet)
+    {
+        if (packet->hasError())
+        {
+            closeLocalTunnel(true, packet->error());
+            return true;
+        }
+        return false;
+    }
+
+    void closeLocalTunnel(bool meet_error, const String & local_err_msg)
+    {
+        std::lock_guard lock(mu);
+        if (!is_done)
+        {
+            is_done = true;
+            consumer_state.setMsg(local_err_msg);
+            local_request_handler.connectionLocalDone(meet_error, local_err_msg);
+        }
+    }
+
+    size_t source_index;
+    LocalRequestHandler local_request_handler;
+
+    std::mutex mu;
+    bool is_done;
```

Comment:
> It looks like it could be replaced with `std::atomic_bool is_done{false};` And then the code of `closeLocalTunnel` would be
> 
> ```
> if (done.compare_exchange_strong(false, true))
> {
>     consumer_state.setMsg(local_err_msg);
>     local_request_handler.connectionLocalDone(meet_error, local_err_msg);
> }
> ```

okk